File: leetcode/2181-merge-nodes-in-between-zeros.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        curr_in = head
        curr_out = head
        while curr_in:
            if curr_in.val == 0:
                # Remove current
                curr_out.next = curr_in.next
                curr_in = curr_in.next
                if curr_in and curr_in.next:
                    curr_in = curr_in.next
                curr_out = curr_out.next
            else:
                # Remove next
                curr_out.val += curr_in.val
                curr_in = curr_in.next


        return head.next

class SolutionSpaceN:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        curr_out = dummy
        curr_in = head
        while curr_in.next:
            if curr_in.val == 0:
                curr_out.next = ListNode()
                curr_out = curr_out.next
            else:
                curr_out.val += curr_in.val

            curr_in = curr_in.next

        return dummy

File: src/2181-merge-nodes-in-between-zeros.js


/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var mergeNodes = function (head) {
    let arr = []
    let curr = head
    let sum = 0
    // go thougth linked list
    while (curr !== null) {
        if (curr.val === 0) {
            arr.push(sum)
            sum = 0
        } else sum += curr.val
        curr = curr.next
    }
    const headResult = new ListNode(arr[1], null)
    let currNodo = headResult
    for (let i = 2; i <= arr.length - 1; i++) {
        let newNode = new ListNode(arr[i], null)
        currNodo.next = newNode
        currNodo = currNodo.next
    }

    return headResult
};